### PR TITLE
Add python solution to usaco: lasers and mirrors

### PR DESCRIPTION
**Pull request by Amaryllidaceae**
_If any of the below don't apply to this Pull Request, mark the checkbox as done._

- [X] I have tested my code.
- [X] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [X] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions), which include the following: 
  - I understand that if it is clear that I have not attempted to follow these guidelines (ex. if I have not used tabs to indent), my PR will be closed.
  - If changes are requested, I will re-request the review after making the changes.

### CODE REVIEW

File: solutions/usaco-671.mdx [hunk 1]
```diff
@@ -84,4 +84,64 @@ int main()
...
 
+<PySection>
+
+```py
+import sys
+from collections import defaultdict, deque
+sys.stdin, sys.stdout = open('lasers.in', 'r'), open('lasers.out', 'w')
+input = sys.stdin.readline
+
+HORIZ, VERTI = 0, 1
```
Comment (SansPapyrus683) [inline solutions/usaco-671.mdx:95]: what do these mean?

File: solutions/usaco-671.mdx [hunk 2]
```diff
@@ -84,4 +84,64 @@ int main()
 
 </CPPSection>
 
+<PySection>
+
+```py
+import sys
+from collections import defaultdict, deque
+sys.stdin, sys.stdout = open('lasers.in', 'r'), open('lasers.out', 'w')
```
Comment (SansPapyrus683) [inline solutions/usaco-671.mdx:92]: assign these on separate lines

File: solutions/usaco-671.mdx [hunk 3]
```diff
@@ -84,4 +84,64 @@ int main()
...
+			if dist[fp_i] == -1:
+				queue.appendleft((fp_i, not curr_direction))
+				dist[fp_i] = dist[curr_idx] + 1
+	else:  # is vertical
+		for fp_i in x_lines[curr_fp.x]:
+			if dist[fp_i] == -1:
+				queue.appendleft((fp_i, not curr_direction))
+				dist[fp_i] = dist[curr_idx] + 1
+
+print(-1)
```
Comment (SansPapyrus683) [inline solutions/usaco-671.mdx:142]: i'd use a `break` and an `else` clause instead of `sys.exit`

File: solutions/usaco-671.mdx (hunk 2 above)
Comment (SansPapyrus683) [reply solutions/usaco-671.mdx:92]: actually i'd rather you use `with` blocks instead of redirecting `stdin` & `stdout`

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/usaco-671.mdx (hunk 2 above)
Comment (Amaryllidaceae) [reply solutions/usaco-671.mdx:92]: Would something like this be fine?
```py
def main(infile, outfile):
	... # solve here

with open('lasers.in') as infile, open('lasers.out') as outfile:
	main(infile, outfile)
```

Comment (SansPapyrus683) [reply solutions/usaco-671.mdx:92]: use 2 separate with blocks

Comment (Amaryllidaceae) [reply solutions/usaco-671.mdx]: not sure if I did it correctly but I did what seemed most reasonable to me

File: solutions/usaco-671.mdx [hunk 4]
```diff
@@ -84,4 +84,63 @@ int main()
...
+<PySection>
+
+```py
+from collections import defaultdict, deque
+
+HORIZ, VERTI = 0, 1  # direction of laser beam (horizontal or vertical)
+
+class Fencepost:
+	def __init__(self, x: int, y: int):
+		self.__slots__ = ['x', 'y']  # faster attribute access
```
Comment (SansPapyrus683) [inline solutions/usaco-671.mdx:96]: does it TLE w/o this?

File: solutions/usaco-671.mdx [hunk 5]
```diff
@@ -84,4 +84,63 @@ int main()
 
 </CPPSection>
 
+<PySection>
+
+```py
+from collections import defaultdict, deque
+
+HORIZ, VERTI = 0, 1  # direction of laser beam (horizontal or vertical)
```
Comment (SansPapyrus683) [inline solutions/usaco-671.mdx:92]: if you're using this to indicate state, might i recommend a boolean or enum instead

File: solutions/usaco-671.mdx [hunk 6]
```diff
@@ -84,4 +84,63 @@ int main()
...
+				queue.appendleft((fp_i, not curr_direction))
+				dist[fp_i] = dist[curr_idx] + 1
+	else:  # is vertical
+		for fp_i in x_lines[curr_fp.x]:
+			if dist[fp_i] == -1:
+				queue.appendleft((fp_i, not curr_direction))
+				dist[fp_i] = dist[curr_idx] + 1
+
+with open('lasers.out', 'w') as outfile:
+	outfile.write(str(output) + '\n')
```
Comment (SansPapyrus683) [inline solutions/usaco-671.mdx:141]:
```suggestion
print(output, file=open('lasers.out', 'w'))
```

File: solutions/usaco-671.mdx [hunk 7]
```diff
@@ -84,4 +84,63 @@ int main()
...
+	for i in range(n):
+		xi, yi = map(int, infile.readline().split())
+		fenceposts.append(Fencepost(xi, yi))
+		x_lines[xi].append(i + 1)
+		y_lines[yi].append(i + 1)
+
+	dist = [-1] * (n + 1)
+	dist[0] = 0
+
+output = -1
```
Comment (SansPapyrus683) [inline solutions/usaco-671.mdx:116]: better name

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/usaco-671.mdx (hunk 4 above)
Comment (Amaryllidaceae) [reply solutions/usaco-671.mdx:96]: nope

File: solutions/usaco-671.mdx [hunk 8]
```diff
@@ -84,4 +84,61 @@ int main()
 
 </CPPSection>
 
+<PySection>
+
+```py
+from collections import defaultdict, deque
+
+HORIZ, VERTI = True, False  # direction of laser beam (horizontal or vertical)
```
Comment (SansPapyrus683) [inline solutions/usaco-671.mdx:92]: i'd ditch the constants here and use them directly
add a comment like "true = horizontal, false = vertical"

**Review by SansPapyrus683 — APPROVED** [review]
yeah sure

**Comment by SansPapyrus683** [comment]
@envyaims review & i'll merge

**Review by envyaims — APPROVED** [review]
seems fine

**Pull request merged**